src/build/env.py

```diff
@@ -12,7 +12,7 @@
 from types import TracebackType
 from typing import Iterable, Optional, Tuple, Type
 
-from ._compat import abstractproperty, add_metaclass
+from ._compat import add_metaclass
```

Comment:
Eh, why revert this part. I prefer the way it was previously 🤔

Reply:
It was breaking mypy.

Reply:
Wait, do you have an old mypy. It worked in the CI and locally for me?

Reply:
Check here https://github.com/pypa/build/runs/1293111599#step:6:7, it passes 👍

Reply:
```
$ git checkout 6efa710
$ mypy src/build
src/build/__main__.py:45: error: Incompatible types in assignment (expression has type "Callable[[], str]", variable has type "Union[bytes, str]")
Found 1 error in 1 file (checked 5 source files)
```

Reply:
eh, ``mypy --version``?

Reply:
```
$ mypy --version
mypy 0.790
```

The CI also failed.

Reply:
Smaller fix for that that keeps the compatibility layer for that separate:

```
--- a/src/build/_compat.py
+++ b/src/build/_compat.py
@@ -32,8 +32,8 @@ else:

     F = TypeVar('F', bound=Callable[..., Any])

-    def abstractproperty(func):  # type: (F) -> F
-        return cast(F, property(abc.abstractmethod(func)))
+    def abstractproperty(func):  # type: (F) -> Any
+        return cast(Any, property(abc.abstractmethod(func)))

```

Reply:
Returning `Any` will result in loss of type information.  The issue here is that mypy cannot resolve an `abstractproperty` function into a `property`, even if you were to return a simple `property`.  I suggest:

```diff
diff --git a/src/build/_compat.py b/src/build/_compat.py
index 3a5da79..ab95485 100644
--- a/src/build/_compat.py
+++ b/src/build/_compat.py
@@ -1,6 +1,6 @@
 import abc
 import sys
-from typing import Callable, TypeVar
+from typing import TYPE_CHECKING, Callable, TypeVar
 
 _T = TypeVar('_T')
 
@@ -28,12 +28,12 @@ def add_metaclass(metaclass):  # type: (type) -> Callable[[_T], _T]
 if sys.version_info[0] == 2:
     abstractproperty = abc.abstractproperty
 else:
-    from typing import Any, cast
+    if TYPE_CHECKING:
+        abstractproperty = abc.abstractproperty
+    else:
 
-    F = TypeVar('F', bound=Callable[..., Any])
-
-    def abstractproperty(func):  # type: (F) -> F
-        return cast(F, property(abc.abstractmethod(func)))
+        def abstractproperty(func):
+            return property(abc.abstractmethod(func))
 
 
 __all__ = (

```